core/corehttp/gateway_handler.go

```diff
@@ -181,10 +181,17 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 	}
 	originalUrlPath := prefix + requestURI.Path
 
-	// Query parameter handling to support requests produced by navigator.registerProtocolHandler.
-	// E.g. This code will redirect calls to /ipfs/?uri=ipfs%3A%2F%2Fcontent-identifier
-	// to /ipfs/content-identifier.
-	if uri := r.URL.Query().Get("uri"); uri != "" {
+	// ?uri query param support for requests produced by web browsers
+	// via navigator.registerProtocolHandler Web API
+	// https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler
+	// TLDR: redirect /ipfs/?uri=ipfs%3A%2F%2Fcid%3Fquery%3Dval to /ipfs/cid?query=val
+	if uriParam := r.URL.Query().Get("uri"); uriParam != "" {
+		// Browsers will pass URI in URL-escaped form, we need to unescape it first
+		uri, err := url.QueryUnescape(uriParam)
```

Comment:
I think `.Query()` will already decode the query parameters. [From the docs](https://golang.org/pkg/net/url/#URL.Query):

> Query parses RawQuery and returns the corresponding values. It silently discards malformed value pairs. To check errors use ParseQuery.

It's not entirely clear if "parsing" also means "decoding" but the distinction to `RawQuery` and the quick check may be evidence enough 🤷‍♂️:

```go
u, _ := url.Parse("http://example.com/?uri=ipfs%3A%2F%2F")
q := u.Query()
fmt.Println(q.Get("uri")) 
// ipfs://
```

Reply:
@dennis-tra good catch! 
I think I was exploring API in a go playground with snippet that did not use `.Query` and did not check if it's truly needed in our case.  Removed in d4ce4b1